### PR TITLE
Reuse WriteBatch

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
## What have you changed? (mandatory)

- Reuse WriteBatch in apply worker
- Reuse WriteBatch in raftstore
- Reuse WriteBatch in apply plain snapshot
- Reuse WriteBatch in batch deletions.

## What are the type of the changes? (mandatory)

- Improvement (change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

Unit test

## Does this PR affect documentation (docs) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

## Refer to a related PR or issue link (optional)

https://github.com/tikv/tikv/issues/4435

## Benchmark result if necessary (optional)

- [x] benchmark

Insert 5,000,000 rows and print jemalloc statistics:

memory size | master nrequests | reuse-wb nrequest
-- | -- | --
4096 | 11572079 | 11619882
8192 | 3076060 | 1398918
16384 | 399087 | 13390
32768 | 66902 | 2109
65536 | 355976 | 329180

From above statistics, we can see this change reduce 97% memory allocations for size of 16K/32K, and reduce 55% memory allocations for size of 8K.

### CODE REVIEW

File: src/raftstore/store/fsm/store.rs [hunk 1]
```diff
@@ -468,30 +470,37 @@ impl<T: Transport, C: PdClient> RaftPoller<T, C> {
...
                 .engines
                 .kv
-                .write_opt(kv_wb, &write_opts)
+                .write_opt(&self.poll_ctx.kv_wb, &write_opts)
                 .unwrap_or_else(|e| {
                     panic!("{} failed to save append state result: {:?}", self.tag, e);
                 });
+            let data_size = self.poll_ctx.kv_wb.data_size();
+            if data_size > KV_WB_SHRINK_SIZE {
+                self.poll_ctx.kv_wb = WriteBatch::new();
```
Comment (Connor1996) [inline src/raftstore/store/fsm/store.rs]: why not with capacity

File: src/raftstore/store/snap.rs [hunk 2]
```diff
@@ -2493,7 +2493,7 @@ pub mod tests {
             .unwrap();
 
             // TODO: this size may change in different RocksDB version.
-            let snap_size = 1342;
+            let snap_size = 1438;
```
Comment (Connor1996) [inline src/raftstore/store/snap.rs]: better use const instead of a magic number? and how to calculate it?

Comment (zhangjinpeng1987) [reply src/raftstore/store/snap.rs]: It's depends on rocksdb's sst format. I think it is ok in test.

**Pull request merged**